src/core/quantum_gate.jl

```diff
@@ -101,6 +133,106 @@ get_connected_qubits(gate::AbstractGate)=[gate.target]
+
+The `ControlledGate` object allows the construction of a controlled `Gate` using a 
+single-target `Operator` (the `kernel`) and the corresponding control and target qubits, 
+specified as `[control_qubit,target_qubit]`.
+
+# Examples
+Whereas a `Hadamard` Gate with `target=1` is constructed by calling `hadamard(target)`, a 
+controlled Hadamard gate is constructed using the `hadamard()` `Function`, and by specifying
+a `connected_qubits=[control, target]`. For instance, the construction 
+of a `ControlledGate{Hadamard}` with `control=1` and `target=2` is performed by calling:
+
+```jldoctest controlled_hadamard
+julia> controlled_hadamard=ControlledGate(hadamard,[1,2])
+Gate Object: ControlledGate{Snowflake.Hadamard}
+Connected_qubits	: [1, 2]
+Operator:
+(4, 4)-element Snowflake.DenseOperator:
+Underlying data ComplexF64:
+1.0 + 0.0im    0.0 + 0.0im    0.0 + 0.0im    0.0 + 0.0im
+0.0 + 0.0im    1.0 + 0.0im    0.0 + 0.0im    0.0 + 0.0im
+0.0 + 0.0im    0.0 + 0.0im    0.7071067811865475 + 0.0im    0.7071067811865475 + 0.0im
+0.0 + 0.0im    0.0 + 0.0im    0.7071067811865475 + 0.0im    -0.7071067811865475 + 0.0im
+
+```
+
+It can then be used in a `QuantumCircuit` as any other `Gate`, and its display symbol is 
+inherited from the display symbol of the single-target `Hadamard` `Gate`:
+
+```jldoctest controlled_hadamard
+julia> circuit=QuantumCircuit(qubit_count=2,gates=[controlled_hadamard])
+Quantum Circuit Object:
+   qubit_count: 2 
+q[1]:──*──
+       |  
+q[2]:──H──
+          
+```
+
+
+"""
+struct ControlledGate{G<:AbstractGate}<:AbstractControlledGate
+    kernel::AbstractOperator
+    connected_qubits::Vector{Int}
+
+    function ControlledGate(
+        kernel_handle::Function,
+        connected_qubits::Vector{Int};
+        params::Vector{<:Real}=Vector{Float64}(),
+        precision::Integer=4)
+
+        @assert length(connected_qubits)==2 "Not implemented for multi-control kernel Operators"
+
+        @assert connected_qubits[1]!=connected_qubits[2] "Control and target qubit must be different, received $connected_qubits"
+
+        @assert haskey(operator_to_gate_mapping,kernel_handle) "Function $kernel_handle is not associated with a Gate constructor"
+
+        # calls function, returns Operator
+        kernel=kernel_handle(params...) 
+
+        # Gate type associated with this operator is used to distinguish between 
+        # ControlledGate types and to assign a display symbol to them
+        kernel_gate_type=operator_to_gate_mapping[kernel_handle]  
+
+        # build new display symbol using existing symbol pertaining to kernel
+
+        symbol_specs=gates_display_symbols[kernel_gate_type]      
+        num_params=length(params)
+        gate_params=Dict{String,Real}() 
+        
+        for (field,param) in zip(symbol_specs[end-num_params+1:end],params)
+            gate_params[field]=param
+        end
+
+        gates_display_symbols[ControlledGate{kernel_gate_type}]=[
```

Comment:
Mutating `gates_display_symbols` is hard to follow and non-obvious for someone not familiar with Snowflake.

If someone wanted to update how the `ControlledGate{XGate}` rendered, for example. They would look at `Base.show` they would, follow along and eventually see that it looks inside `gates_display_symbols`, and would be very confused to see that `ControlledGate{XGate}` is not present. They might find this if they search for all places `gates_display_symbols` is defined, but this is not obvious, and mutable changes like this tend to become fragile over time.

A good mental check to do to verify if code is readable is to put yourself in the shoes of someone that has never seen this code before. How will they navigate this code?

We need to improve this at some point, but for now, let's simply have a `get_controlled_gate_symbol` function which looks at the inner gate and adds the `*` symbols.

Reply:
@plerouxanyon do you mean that display symbol construction will be handled in `get_display_symbol(::ControlledGate)` and not stored in `gates_display_symbols`?

Reply:
Yes, since all controlled gates will be handled the same way. The controlled qubits will all be `*`, and the rest will be the regular symbols from `get_display_symbols`.

Reply:
updated